aws-observability/aws-application-signals-test-framework
# Ec2 cleanup workflow and script

### Description of changes:
- Create a workflow that runs every day at 18:00 UTC (11AM PST) to perform clean up of AWS resources in the E2E testing accounts in all regions. The workflow can also be run on demand if needed.
- For now the workflow only cleans up the EC2 resources that are:
  - more than 3 hrs old at the time the workflow runs
  - do NOT belong to an EKS cluster (by checking their security group name)
  - do NOT belong to a K8s cluster (by checking the key name)
  - do NOT have the `do-not-delete` tag.
- It is possible that we may unintentionally terminate some hosts. The cleanup job will upload the full report of instances filtered for termination and upload to an S3 bucket so that we can audit them and fix/improve our filter logic.

NOTE: I will submit another PR to uncomment the termination logic.

### Testing
- Test run: https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/9913944002/job/27392017400
  - The bucket name was intentionally incorrect since it is a WIP.
  - The termination part was commented out intentionally. We will process terminations once this PR is reviewed and merged.  

--- 
*Ensure you've run the following tests on your changes and include the link below:*
To do so, create a `test.yml` file with `name: Test` and workflow description to test your changes, then remove the file for your PR. Link your test run in your PR description. This process is a short term solution while we work on creating a staging environment for testing.

NOTE: TESTS RUNNING ON A SINGLE EKS CLUSTER CANNOT BE RUN IN PARALLEL. See the [needs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds) keyword to run tests in succession.
- Run Java EKS on `e2e-playground` in us-east-1 and eu-central-2
- Run Python EKS on `e2e-playground` in us-east-1 and eu-central-2
- Run metric limiter on EKS cluster `e2e-playground` in us-east-1 and eu-central-2
- Run EC2 tests in all regions
- Run K8s on a separate K8s cluster (check IAD test account for master node endpoints; these will change as we create and destroy clusters for OS patching)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
